source/particles/particle_handler.cc

```diff
@@ -1222,140 +1205,141 @@ namespace Particles
-    const void *pdata =
-      reinterpret_cast<const void *>(n_particles_in_cell_ptr + 1);
-
-    if (*n_particles_in_cell_ptr == 0)
-      return;
+    std::vector<Particle<dim, spacedim>> loaded_particles_on_cell =
+      Utilities::unpack<std::vector<Particle<dim, spacedim>>>(
+        data_range.begin(),
+        data_range.end(),
+        /*allow_compression=*/true);
 
-    // Load all particles from the data stream and store them in the local
-    // particle map.
-    if (status ==
-        parallel::distributed::Triangulation<dim, spacedim>::CELL_PERSIST)
+    switch (status)
       {
-        typename std::multimap<internal::LevelInd,
-                               Particle<dim, spacedim>>::iterator
-          position_hint = particles.end();
-        for (unsigned int i = 0; i < *n_particles_in_cell_ptr; ++i)
+        case parallel::distributed::Triangulation<dim, spacedim>::CELL_PERSIST:
           {
-            // Use std::multimap::emplace_hint to speed up insertion of
-            // particles. This is a C++11 function, but not all compilers
-            // that report a -std=c++11 (like gcc 4.6) implement it, so
-            // require C++14 instead.
+            typename std::multimap<internal::LevelInd,
+                                   Particle<dim, spacedim>>::iterator
+              position_hint = particles.end();
+            for (auto &particle : loaded_particles_on_cell)
+              {
+                // Use std::multimap::emplace_hint to speed up insertion of
+                // particles. This is a C++11 function, but not all compilers
+                // that report a -std=c++11 (like gcc 4.6) implement it, so
+                // require C++14 instead.
 #  ifdef DEAL_II_WITH_CXX14
-            position_hint = particles.emplace_hint(
-              position_hint,
-              std::make_pair(cell->level(), cell->index()),
-              Particle<dim, spacedim>(pdata, property_pool.get()));
+                position_hint =
+                  particles.emplace_hint(position_hint,
+                                         std::make_pair(cell->level(),
+                                                        cell->index()),
+                                         std::move(particle));
 #  else
-            position_hint = particles.insert(
-              position_hint,
-              std::make_pair(std::make_pair(cell->level(), cell->index()),
-                             Particle<dim, spacedim>(pdata,
-                                                     property_pool.get())));
+                position_hint =
+                  particles.insert(position_hint,
+                                   std::make_pair(std::make_pair(cell->level(),
+                                                                 cell->index()),
+                                                  std::move(particle)));
 #  endif
-            ++position_hint;
+                ++position_hint;
```

Comment:
add a comment here like "move the hint position forward by one, i.e., for the next particle, the 'hint' position will be right after the one just inserted" or similar